runtime/precompiles/src/lib.rs

```diff
@@ -16,35 +16,66 @@
 
 #![cfg_attr(not(feature = "std"), no_std)]
 
+mod staking;
 use codec::Decode;
 use evm::{Context, ExitError, ExitSucceed};
 use frame_support::dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo};
-use pallet_evm::{Config, Precompile, PrecompileSet};
+use pallet_evm::{Precompile, PrecompileSet};
 use pallet_evm_precompile_bn128::{Bn128Add, Bn128Mul, Bn128Pairing};
 use pallet_evm_precompile_dispatch::Dispatch;
 use pallet_evm_precompile_modexp::Modexp;
 use pallet_evm_precompile_sha3fips::Sha3FIPS256;
 use pallet_evm_precompile_simple::{ECRecover, Identity, Ripemd160, Sha256};
 use sp_core::H160;
+use sp_std::convert::TryFrom;
+use sp_std::fmt::Debug;
 use sp_std::{marker::PhantomData, vec::Vec};
+use staking::ParachainStakingWrapper;
 
+use frame_support::traits::Currency;
+type BalanceOf<Runtime> = <<Runtime as parachain_staking::Config>::Currency as Currency<
+	<Runtime as frame_system::Config>::AccountId,
+>>::Balance;
+
+//TODO Maybe we don't need to / shouldn't be generic over the runtime.
+// Pros: Would simplify trait bounds and speed up compile time (maybe not noticeably).
+// Cons: Would proclude using this precompile set in mocked Runtimes.
 /// The PrecompileSet installed in the Moonbeam runtime.
 /// We include the nine Istanbul precompiles
 /// (https://github.com/ethereum/go-ethereum/blob/3c46f557/core/vm/contracts.go#L69)
 /// as well as a special precompile for dispatching Substrate extrinsics
 #[derive(Debug, Clone, Copy)]
 pub struct MoonbeamPrecompiles<R>(PhantomData<R>);
 
+// The idea here is that we won't have to list the addresses in this file and the chain spec.
+// Unfortunately we still have to type it twice in this file.
+impl<R: frame_system::Config> MoonbeamPrecompiles<R>
+where
+	R::AccountId: From<H160>,
+{
+	/// Return all addresses that contain precompiles. This can be used to populate dummy code
+	/// under the precompile, and potentially in the future to prevent using accounts that have
+	/// precompiles at their addresses explicitly using something like SignedExtra.
+	#[allow(dead_code)]
+	fn used_addresses() -> impl Iterator<Item = R::AccountId> {
+		sp_std::vec![1, 2, 3, 4, 5, 6, 7, 8, 255, 256, 511]
+			.into_iter()
+			.map(|x| hash(x).into())
+	}
```

Comment:
Just make sure there is a followup issue for this. Then I'm fine leaving it as it is.

Reply:
These addresses have been updated in `master`, by the way. We pushed some out to `1024` and I (somewhat arbitrarily) placed `sacrifice` at `4095`.